--- Makefile.bak	2017-11-06 14:38:19.663256474 +0000
+++ Makefile	2017-11-06 14:39:23.357735899 +0000
@@ -1169,7 +1169,7 @@
 # relocation).
 # FIXME refactor dts/Makefile to share target/arch detection
 u-boot.elf: u-boot.bin
-	@$(OBJCOPY)  -B aarch64 -I binary -O elf64-littleaarch64 \
+	@$(OBJCOPY)  -B arm -I binary -O elf32-littlearm \
 		$< u-boot-elf.o
 	@$(LD) u-boot-elf.o -o $@ \
 		--defsym=_start=$(CONFIG_SYS_TEXT_BASE) \