diff --git a/hw/misc/zynq_slcr.c b/hw/misc/zynq_slcr.c
index 7557482..3a387ec 100644
--- a/hw/misc/zynq_slcr.c
+++ b/hw/misc/zynq_slcr.c
@@ -310,7 +310,7 @@ static void zynq_slcr_reset(DeviceState *d)
 
     s->regs[R_LOCKSTA] = 1;
     /* 0x100 - 0x11C */
-    s->regs[R_ARM_PLL_CTRL]   = 0x0001A008;
+    s->regs[R_ARM_PLL_CTRL]   = 0x00014008;
     s->regs[R_DDR_PLL_CTRL]   = 0x0001A008;
     s->regs[R_IO_PLL_CTRL]    = 0x0001A008;
     s->regs[R_PLL_STATUS]     = 0x0000003F;
